src/primitives/transaction.h

```diff
@@ -195,6 +198,59 @@ class CTxOut
 
 struct CMutableTransaction;
 
+template<typename Stream, typename TxType>
+inline void SerializeTransaction(TxType& tx, Stream& s, int nType, int nVersion) {
+    ser_writedata32(s, tx.nVersion);
+    nVersion = tx.nVersion;
+    if (nVersion >= 4) {
+        for (auto in : tx.vin) {
+            CMFToken hash(Consensus::TxInPrevHash, in.prevout.hash);
+            STORECMF(hash);
+            if (in.prevout.n > 0) {
+                CMFToken index(Consensus::TxInPrevIndex, (uint64_t) in.prevout.n);
+                STORECMF(index);
+            }
+            // TODO check sequence to maybe store the BIP68 stuff.
+        }
+        for (auto out : tx.vout) {
+            CMFToken token(Consensus::TxOutValue, (uint64_t) out.nValue);
+            STORECMF(token);
+            std::vector<char> script(out.scriptPubKey.begin(), out.scriptPubKey.end());
+            token = CMFToken(Consensus::TxOutValue, script);
+            STORECMF(token);
+        }
+        for (auto in : tx.vin) {
+            CMFToken token(Consensus::TxInScript, std::vector<char>(in.scriptSig.begin(), in.scriptSig.end()));
+            STORECMF(token);
+        }
+        CMFToken end(Consensus::TxEnd);
+        STORECMF(end);
+    } else {
+        CSerActionSerialize ser_action;
+        READWRITE(tx.vin);
+        READWRITE(tx.vout);
+        READWRITE(tx.nLockTime);
+    }
+}
+
+void loadTransaction(const std::vector<CMFToken>& tokens, std::vector<CTxIn> &inputs, std::vector<CTxOut> &outputs);
+
+template<typename Stream, typename TxType>
+inline void UnSerializeTransaction(TxType& tx, Stream& s, int nType, int nVersion) {
+    *const_cast<int32_t*>(&tx.nVersion) = ser_readdata32(s);
+    nVersion = tx.nVersion;
+    if (nVersion >= 4) {
+        loadTransaction(UnserializeCMFs(s, Consensus::TxEnd, nType, nVersion),
+            *const_cast<std::vector<CTxIn>*>(&tx.vin),
+            *const_cast<std::vector<CTxOut>*>(&tx.vout));
+    } else {
+        CSerActionUnserialize ser_action;
+        READWRITE(*const_cast<std::vector<CTxIn>*>(&tx.vin));
+        READWRITE(*const_cast<std::vector<CTxOut>*>(&tx.vout));
+        READWRITE(*const_cast<uint32_t*>(&tx.nLockTime));
+    }
+}
+
```

Comment:
This is historical; the CTransaction uses public const members for whatever stupid reason. Its also the reason there is a CMutableTransaction.